lms/djangoapps/courseware/views/views.py

```diff
@@ -1513,3 +1515,36 @@ def financial_assistance_form(request):
             }
         ],
     })
+
+
+@transaction.non_atomic_requests
+@require_POST
+def end_course(request):
+    params, response = _validate_post_params(request.POST)
+    if response is not None:
+        return response
+
+    try:
+        # Check that the course exists
```

Comment:
I would use the next method `CourseEnrollment.get_enrollment`: https://github.com/raccoongang/edx-platform/blob/add_button_end_course/common/djangoapps/student/models.py#L1066 
And  I would save it to the variable (`enrollment_state` as example). After that, I would check the `enrollment_state.is_active if method returned the CourseEnrollment object`. If everything OK,  I would run `generate_certificates_for_students` and run `reverse`.
I think this takes less code and fewer queries to the database.